lib/active_record_view_model/update_data.rb

```diff
@@ -32,53 +56,110 @@ class UpdateData
+          'description'          => 'viewmodel reference',
+          'properties'           => { TYPE_ATTRIBUTE => type,
+                                      ID_ATTRIBUTE   => id },
+          'additionalProperties' => false,
+          'required'             => [TYPE_ATTRIBUTE, ID_ATTRIBUTE]
+        }
+
+      base_functional_update_schema =
+        {
+          'type'                 => 'object',
+          'properties'           => {
+            VALUES_ATTRIBUTE => { 'type' => 'array' }
           },
-          'required'    => [TYPE_ATTRIBUTE, VALUES_ATTRIBUTE],
+          'additionalProperties' => false,
+          'required'             => [TYPE_ATTRIBUTE, VALUES_ATTRIBUTE]
         }
 
+      append = base_functional_update_schema.deep_merge(
+        {
+          'description' => 'collection append',
+          'properties'  => {
+            TYPE_ATTRIBUTE   => { 'enum' => [FunctionalUpdate::Type::Append.name.downcase] },
+            VALUES_ATTRIBUTE => { 'items' => viewmodel_update },
+            BEFORE_ATTRIBUTE => viewmodel_reference,
+            AFTER_ATTRIBUTE  => viewmodel_reference
+          }
+        }
+      )
+      APPEND_ACTION = JsonSchema.parse!(append)
+
+      update = base_functional_update_schema.deep_merge(
+        {
+          'description' => 'collection update',
+          'properties'  => {
+            TYPE_ATTRIBUTE   => { 'enum' => [FunctionalUpdate::Type::Update.name.downcase] },
+            VALUES_ATTRIBUTE => { 'items' => viewmodel_update },
+          }
+        }
+      )
+      UPDATE_ACTION = JsonSchema.parse!(update)
+
+      remove = base_functional_update_schema.deep_merge(
+        {
+          'description' => 'collection remove',
+          'properties'  => {
+            TYPE_ATTRIBUTE   => { 'enum' => [FunctionalUpdate::Type::Remove.name.downcase] },
+            VALUES_ATTRIBUTE => { 'items' => viewmodel_update },
+            # The VALUES_ATTRIBUTE should be a viewmodel_reference, but in the
+            # name of error messages, we allow more keys and check the
+            # constraint in code.
+          }
+        }
+      )
+      REMOVE_ACTION = JsonSchema.parse!(remove)
+
       collection_update =
         {
           'type'        => 'object',
           'description' => 'collection functional update',
           'properties'  => {
             TYPE_ATTRIBUTE    => { 'enum' => [FUNCTIONAL_UPDATE_TYPE] },
-            ACTIONS_ATTRIBUTE => { 'type'  => 'array',
-                                   'items' => collection_update_action }
+            ACTIONS_ATTRIBUTE => { 'type'  => 'array' }
+            # The ACTIONS_ATTRIBUTE could be accurately expressed as
+            #
+            #   { 'oneOf' => [append, update, remove] }
+            #
+            # but this produces completely unusable error messages.  Instead we
+            # specify it must be an array, and defer checking to the code that
+            # can determine the schema by inspecting the type field.
```

Comment:
In that case could you at least check here that it's a collection update base (allowing additional properties), and like you say do the subtype checking individually after match?